pkg/controllers/k8sdefaultendpoint/k8sdefaultendpoint.go

```diff
@@ -74,51 +80,155 @@ func syncKubernetesServiceEndpoints(ctx context.Context, virtualClient client.Cl
 		return err
 	}
 
-	// get virtual service endpoints
-	vObj := &corev1.Endpoints{}
-	err = virtualClient.Get(ctx, types.NamespacedName{
-		Namespace: "default",
-		Name:      "kubernetes",
-	}, vObj)
-	if err != nil {
-		if kerrors.IsNotFound(err) {
+	vEndpoints := &corev1.Endpoints{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "default",
+			Name:      "kubernetes",
+		},
+	}
+
+	result, err := controllerutil.CreateOrPatch(ctx, virtualClient, vEndpoints, func() error {
+		if vEndpoints.Labels == nil {
+			vEndpoints.Labels = map[string]string{}
+		}
+		vEndpoints.Labels[discovery.LabelSkipMirror] = "true"
+
+		// build new subsets
+		newSubsets := pEndpoints.DeepCopy().Subsets
+		for i := range newSubsets {
+			for j := range newSubsets[i].Ports {
+				newSubsets[i].Ports[j].Name = "https"
+			}
+			for j := range pEndpoints.Subsets[i].Addresses {
+				newSubsets[i].Addresses[j].Hostname = ""
+				newSubsets[i].Addresses[j].NodeName = nil
+				newSubsets[i].Addresses[j].TargetRef = nil
+			}
+			for j := range pEndpoints.Subsets[i].NotReadyAddresses {
+				newSubsets[i].NotReadyAddresses[j].Hostname = ""
+				newSubsets[i].NotReadyAddresses[j].NodeName = nil
+				newSubsets[i].NotReadyAddresses[j].TargetRef = nil
+			}
+		}
+
+		oldJSON, err := json.Marshal(vEndpoints.Subsets)
+		if err != nil {
+			return err
+		}
+		newJSON, err := json.Marshal(newSubsets)
+		if err != nil {
+			return err
+		}
+
+		if string(oldJSON) == string(newJSON) {
 			return nil
 		}
 
+		vEndpoints.Subsets = newSubsets
+
+		return nil
+	})
+	if err != nil {
+		return nil
+	}
+
+	if result == controllerutil.OperationResultCreated || result == controllerutil.OperationResultUpdated {
+		vSlices := &discovery.EndpointSlice{
```

Comment:
Why do we need to do that manually instead of using the mirror controller from k8s?

Reply:
I am just doing exactly what k8s does - https://github.com/kubernetes/kubernetes/blob/7380fc735aca591325ae1fabf8dab194b40367de/pkg/controlplane/reconcilers/lease.go#L200-L207
the mirror controller would probably work, but I am not 100% sure, so I went with this as a safer option.